File: 0114_Flatten Binary Tree to Linked List/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public void flatten(TreeNode root) {
        TreeNode current = root;
        while (current != null) {
            // If the node has a left subtree
            if (current.left != null) {
                // Find the rightmost node of the left subtree
                TreeNode rightmost = current.left;
                while (rightmost.right != null) {
                    rightmost = rightmost.right;
                }
                // Reconnect the rightmost node of the left subtree to the right subtree
                rightmost.right = current.right;
                // Move the left subtree to the right
                current.right = current.left;
                current.left = null;
            }
            // Move to the next node (right child)
            current = current.right;
        }
    }
}

public class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();

        // Example 1: Tree [1, 2, 5, 3, 4, null, 6]
        TreeNode root1 = new TreeNode(1);
        root1.left = new TreeNode(2);
        root1.right = new TreeNode(5);
        root1.left.left = new TreeNode(3);
        root1.left.right = new TreeNode(4);
        root1.right.right = new TreeNode(6);

        solution.flatten(root1);
        printFlattenedTree(root1); // Output: [1, null, 2, null, 3, null, 4, null, 5, null, 6]

        // Example 2: Tree []
        TreeNode root2 = null;
        solution.flatten(root2);
        printFlattenedTree(root2); // Output: []

        // Example 3: Tree [0]
        TreeNode root3 = new TreeNode(0);
        solution.flatten(root3);
        printFlattenedTree(root3); // Output: [0]
    }

    private static void printFlattenedTree(TreeNode root) {
        TreeNode current = root;
        while (current != null) {
            System.out.print(current.val + " ");
            current = current.right;
        }
        System.out.println();
    }
}

File: 0114_Flatten Binary Tree to Linked List/Solution.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def flatten(self, root):
        current = root
        while current:
            if current.left:
                # Find the rightmost node of the left subtree
                rightmost = current.left
                while rightmost.right:
                    rightmost = rightmost.right
                # Connect the rightmost node of the left subtree to the current node's right subtree
                rightmost.right = current.right
                # Move the left subtree to the right and set the left pointer to None
                current.right = current.left
                current.left = None
            # Move to the next node (right child)
            current = current.right

# Helper function to create a tree from a list
def create_tree_from_list(values):
    if not values:
        return None
    nodes = [TreeNode(val) if val is not None else None for val in values]
    for i in range(len(nodes)):
        if nodes[i] is not None:
            if 2 * i + 1 < len(nodes):
                nodes[i].left = nodes[2 * i + 1]
            if 2 * i + 2 < len(nodes):
                nodes[i].right = nodes[2 * i + 2]
    return nodes[0]

# Helper function to print the flattened tree as a list
def print_flattened_tree(root):
    current = root
    result = []
    while current:
        result.append(current.val)
        current = current.right
    print(result)

# Example usage
solution = Solution()

# Example 1: Tree [1, 2, 5, 3, 4, None, 6]
root1 = create_tree_from_list([1, 2, 5, 3, 4, None, 6])
solution.flatten(root1)
print_flattened_tree(root1)  # Output: [1, 2, 3, 4, 5, 6]

# Example 2: Tree []
root2 = create_tree_from_list([])
solution.flatten(root2)
print_flattened_tree(root2)  # Output: []

# Example 3: Tree [0]
root3 = create_tree_from_list([0])
solution.flatten(root3)
print_flattened_tree(root3)  # Output: [0]
